### PR TITLE
Expand bone name possibilities.

### LINKED ISSUE #1855: Add option to opt in to bone name sanitization in GLTF importer
**Describe the project you are working on:**
It's a 3D character creator game with skeletal body customization and wearables

**Describe the problem or limitation you are having in your project:**
We rely on accessing the character skeleton programmatically to customize the body but the skeleton bones names have been changed from what was originally exported as GLTF from Blender. Our original bones use the "a01_S_BoneName_01" naming pattern but it is changed to a snake_case form on import by the GLTF importer to something like "a01_s_bone_name_1" thus making access to bones unpredictable.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
While the bone name sanitization process the GLTF importer does surely solves a corner case, it shouldn't be applied to every GLTF document. Skipping the renaming of bones lets users know their bone names remain unchanged and allows them to keep their preferred naming pattern. An import option should be added to the importer to opt in to this feature.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
A boolean option named, for example, "verbatim names" is added to the importer. If enabled, the bone name sanitization is skipped on import/re-import.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
The features discussed here occur during the import process and can't be worked around with a script.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This is a change of behavior proposal for a core module.

Originally reported as godotengine/godot#43482

### LINKED ISSUE #43482: Unexpected bone names because GLTF importer renames bones
**Godot version:**
3.2.3

**OS/device including version:**
Debian unstable

**Issue description:**
The GLTF importer renames bones on import in a process called "sanitization". It works by converting the bone name to some snake_case pattern that also muddles numbers. It probably solves a corner case but I don't think it should be applied every time. The problem is that I have to access the skeleton bones programmatically and the importer gets in the way by altering the data (bone names) I know. 

I think an import option should be added to opt in to the bone name sanitization process.

**Steps to reproduce:**
* Run the reproduction project and move the slider. It will fail because the expected bone name in the script isn't the actual imported bone name.

**Minimal reproduction project:**
[GltfImporterRenamesBonese.zip](https://github.com/godotengine/godot/files/5533083/GltfImporterRenamesBonese.zip)

### PR DESCRIPTION
**Pull request by fire**
Reduce the limits on bone names.

Continues: https://github.com/godotengine/godot/pull/46939/files
Fixes: https://github.com/godotengine/godot/issues/43482
Fixes: https://github.com/godotengine/godot-proposals/issues/1855

### CODE REVIEW

File: modules/gltf/gltf_document.cpp [hunk 1]
```diff
@@ -505,25 +505,8 @@ String GLTFDocument::_gen_unique_animation_name(Ref<GLTFState> state, const Stri
...
-	p_name = pattern_noslash.sub(p_name, "_", true);
-
-	RegEx pattern_nospace(" +");
-	p_name = pattern_nospace.sub(p_name, "_", true);
-
-	RegEx pattern_multiple("_+");
-	p_name = pattern_multiple.sub(p_name, "_", true);
-
-	RegEx pattern_padded("0+(\\d+)");
-	p_name = pattern_padded.sub(p_name, "$1", true);
```
Comment (fire) [inline modules/gltf/gltf_document.cpp:524]: ": /" are already banned.

` ` Spaces are valid.

`_`  Multiple underscores is valid.

Regex "0+(\\d+)" appears to be a way hide injected nodes.

Comment (fire) [reply modules/gltf/gltf_document.cpp:524]: ". : @ / \""

". @ \" I'm pretty sure these are all valid skeleton characters.

Comment (fire) [reply modules/gltf/gltf_document.cpp:524]: ". : @ / "" are valid Skeleton bone names.

image

Comment (fire) [reply modules/gltf/gltf_document.cpp]: image

The minimum appears to be stripping ": /"

**Review by reduz — APPROVED** [review]

**Comment by fire** [comment]
Updated test project.

[GltfImporterRenamesBoneseUpdated.zip](https://github.com/godotengine/godot/files/6151907/GltfImporterRenamesBoneseUpdated.zip)

**Comment by fire** [comment]
I noticed something weird where I lose the animation track. Doing another testing pass.

**Comment by fire** [comment]
I took https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/RiggedSimple/glTF-Binary/RiggedSimple.glb and renamed the bones to `.:@/"`.

The `:/` character were stripped and the animation still work.

**Comment by akien-mga** [comment]
Since there are only two trivial patterns, maybe you could use `String::replace()` instead of RegEx and remove the RegEx dependency in this code? It's a module that can be disabled in theory, and this dependency was not properly guarded with `MODULE_REGEX_ENABLED`, but now we can simply remove it. (That's not critical though, just a suggestion.)

And so for bones we don't need to apply the same replaces as in `String::validate_node_name()`?

**Comment by fire** [comment]
I'll use replace.

Also, they never used the same replaces as `String::validate_node_name()` so I'm not going to limit them in the revision.

**Review by akien-mga — COMMENTED** [review]
Looks good, please also remove the `#include` for the RegEx module.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.3.

**Pull request merged**